### PR TITLE
Async rework

### PR DESCRIPTION
**Pull request by kevmoo**
CC @kwalrath @sethladd 

More to do, but this is a start

### CODE REVIEW

**Comment by kevmoo** [comment]
Can I get a review? @kwalrath @sethladd @Sfshaza

File: lib/dartiverse_search.dart [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * The WebSocket can send search requests as JSON-formatted messages,
+ * which will be responded to with a series of results and finally a done
+ * message.
+ */
+void handleWebSocket(WebSocket webSocket) {
+  log.info('New WebSocket connection');
+
+  // Listen for incoming data. We expect the data to be a JSON-encoded String.
+  webSocket.map((string) => JSON.decode(string)).listen((json) {
```
Comment (kwalrath) [inline lib/dartiverse_search.dart]: Should this use await for?

File: bin/server.dart [hunk 2]
```diff
@@ -93,18 +24,19 @@ void main() {
...
 
     // The client will connect using a WebSocket. Upgrade requests to '/ws' and
     // forward them to 'handleWebSocket'.
-    router.serve('/ws')
-      .transform(new WebSocketTransformer())
-      .listen(handleWebSocket);
+    router
+        .serve('/ws')
+        .transform(new WebSocketTransformer())
+        .listen(handleWebSocket);
```
Comment (kwalrath) [inline bin/server.dart]: listen -> await for?

File: web/client.dart [hunk 3]
```diff
@@ -6,78 +6,80 @@ import 'dart:async';
...
-    setStatus('');
-    searchElement.disabled = false;
-    searchElement.focus();
-    webSocket.onMessage.listen((e) {
-      handleMessage(e.data);
+  void _onConnected() {
+    _setStatus('');
+    _searchElement.disabled = false;
+    _searchElement.focus();
+    _webSocket.onMessage.listen((e) {
```
Comment (kwalrath) [inline web/client.dart]: use await for?

File: bin/server.dart (hunk 2 above)
Comment (kwalrath) [inline bin/server.dart]: listen() actually seems fine here, but I'm trying to figure out when we should use it vs. await for. listen() seems best for DOM events and cases, in general, where you don't care about waiting for the result.

**Comment by kevmoo** [comment]
@kwalrath `await for` can be used when you want to **block** and wait for the stream to finish.

In the cases above – which I will double check – calls to `listen` shouldn't block the called function. 

A quick way to tell the difference: we don't want to change the return type of the function to be `Future` or `Stream`.

**Comment by kwalrath** [comment]
It looks really good, as far as I can tell, but I didn't try to run it. I like how you split it up into a bunch of targeted commits that are easier to look at than one big one.

**Comment by kwalrath** [comment]
Your explanation sounds good, @kevmoo. I'll use that when I talk about await for.

**Comment by kevmoo** [comment]
@kwalrath it runs great. Verified.

LGTY?

**Comment by kwalrath** [comment]
LGTM!

**Pull request merged**